### PR TITLE
Add new bls operators and an operator compatibility mode

### PR DESCRIPTION
**Pull request by prozacchiwawa**
--operators-version can now be used in situations where disassembly is output to select an older version of the operator set.

### CODE REVIEW

**Comment by prozacchiwawa** [comment]
Gonna add tests of the bls operators then set it Ready.

**Comment by prozacchiwawa** [comment]
ok waiting for green then gonna mark it ready.

**Comment by cameroncooper** [comment]
I think @arvidn wanted to remove the bls prefix from g1 and g2 ops: https://github.com/Chia-Network/clvm_rs/blob/main/op-tests/test-bls-ops.txt#L136-L267

**Comment by prozacchiwawa** [comment]
ahh thanks, hadn't heard anything since i saw the definitions in a previous
version of clvm_rs.
will change them here.

**Comment by freddiecoleman** [comment]
Any chance this could be updated to use clvmr 0.2.6 to get the secp operators as well? Thanks.

**Comment by prozacchiwawa** [comment]
yes, i'll check on what the final opcode for the secp operator turned out
to be.

**Comment by prozacchiwawa** [comment]
ok changed the operator names and added the secp256 operators.  requires a clvmr bump which changes clvmr's api just a bit.

**Comment by freddiecoleman** [comment]
I manually compiled and ran a Chialisp program that uses `secp256r1_verify` and can confirm it returns `()` when run with a valid pk/msg/sig and throws `FAIL: secp256r1_verify failed` when the signature is invalid.

**Comment by prozacchiwawa** [comment]
Thanks for trying it out (and apologies for not commenting sooner).

File: src/classic/clvm_tools/clvmc.rs [hunk 1]
```diff
@@ -33,9 +33,12 @@ fn include_dialect(
     dialects: &HashMap<Vec<u8>, i32>,
     e: &[NodePtr],
 ) -> Option<i32> {
-    if let (SExp::Atom(inc), SExp::Atom(name)) = (allocator.sexp(e[0]), allocator.sexp(e[1])) {
-        if allocator.buf(&inc) == "include".as_bytes().to_vec() {
-            if let Some(dialect) = dialects.get(allocator.buf(&name)) {
+    // Propogated names from let capture to labeled nodes.
+    let inc_node = e[0];
```
Comment (aqk) [inline src/classic/clvm_tools/clvmc.rs]: `inc`?

File: src/classic/clvm_tools/cmds.rs [hunk 2]
```diff
@@ -277,6 +299,17 @@ impl ArgumentValueConv for StageImport {
     }
 }
 
+struct OperatorsVersion {}
+
+impl ArgumentValueConv for OperatorsVersion {
+    fn convert(&self, arg: &str) -> Result<ArgumentValue, String> {
+        let ver = arg
+            .parse::<i64>()
+            .map_err(|_| format!("expected number 0-1 found {arg}"))?;
```
Comment (aqk) [inline src/classic/clvm_tools/cmds.rs]:
```suggestion
            .map_err(|_| format!("expected number 0-1 but found {arg}"))?;
```

File: src/classic/clvm_tools/cmds.rs [hunk 3]
```diff
@@ -772,6 +805,14 @@ fn fix_log(
     }
 }
 
+fn get_disassembly_ver(p: &HashMap<String, ArgumentValue>) -> Option<usize> {
+    if let Some(ArgumentValue::ArgInt(x)) = p.get("operators_version") {
```
Comment (aqk) [inline src/classic/clvm_tools/cmds.rs:809]: CURRENT_OPERATORS_VERSION

File: src/classic/clvm_tools/debug.rs [hunk 4]
```diff
@@ -126,7 +126,7 @@ pub fn build_symbol_dump(
             let args_name_atom = allocator.new_atom(&args_atom)?;
             let left_env_name_atom = allocator.new_atom(&left_env_atom)?;
 
-            let serialized_args = disassemble(allocator, extra.args);
+            let serialized_args = disassemble(allocator, extra.args, Some(0));
```
Comment (aqk) [inline src/classic/clvm_tools/debug.rs:129]: Would be a bit more clear if callsite named version param somehow.

**Comment by aqk** [comment]
Test comment that won't hold up the review.

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-bbc9a97f54d908837e8384bb03c2654d3209f6287dfe6add16deb65562a8720eR2

**Comment by aqk** [comment]
Would be a bit more clear if call site named version param somehow.

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-9fa75347c59434cb17b157cf0cadd4496a701f3d99266b389fc0423d121d089eR129

**Comment by aqk** [comment]
Could you clarify the wording of `and A is bound to B` in this comment? 

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-54c3cf34c36830af104cc1f7fa3318b4c8ff240a50576df76bb61fd172b0a127R48

**Comment by aqk** [comment]
Will this thow if `pright` is not an atom?

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-54c3cf34c36830af104cc1f7fa3318b4c8ff240a50576df76bb61fd172b0a127R64

**Comment by aqk** [comment]
It feels a bit strange to also remove `NO_NEG_DIV` in the same PR.

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-d7bb1fafbfc7e664d696beef72569e9e16c1fcbc9748bd57f292fc10bd7cd9edR2

**Comment by aqk** [comment]
I think you should document the semantics of `OPERATORS_LATEST_VERSION` near the definition. (That the intent is to save the user having to specify the decompile version every time they invoke the tool).

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-619b29147b2c7f544ed77d26e027e880c3b3a8d7fe18f464538a2f8eb06a0ee5R11

**Comment by aqk** [comment]
These two comments about `opbuf` could be more helpful, especially since `opbuf` is no longer bound.

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-56585dbab0da8912246c3ad09e10672b7e24ae2d8246ab236fe774d06b385b36R131

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-56585dbab0da8912246c3ad09e10672b7e24ae2d8246ab236fe774d06b385b36R143

File: src/classic/clvm_tools/stages/stage_2/compile.rs [hunk 5]
```diff
@@ -453,18 +462,18 @@ fn find_symbol_match(
...
                         let symbol = symdef[0];
                         let value = if symdef.len() == 1 {
                             allocator.null()
                         } else {
                             symdef[1]
                         };
 
-                        let symbuf = allocator.buf(&symptr).to_vec();
-                        if vec![b'*'] == symbuf {
+                        let symbuf = allocator.atom(symdef[0]);
```
Comment (aqk) [inline src/classic/clvm_tools/stages/stage_2/compile.rs]:
```suggestion
                        let symbuf = allocator.atom(symbol);
```

**Comment by aqk** [comment]
Much nicer!

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-56585dbab0da8912246c3ad09e10672b7e24ae2d8246ab236fe774d06b385b36R369

**Comment by aqk** [comment]
Is `get_disassembly_ver` used everywhere it could / should?

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-56585dbab0da8912246c3ad09e10672b7e24ae2d8246ab236fe774d06b385b36R779

**Comment by aqk** [comment]
Is removing `val_stack_limit` connected with this change? If not, it could be a separate PR.

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-2074b76a9ca8bff0a579d5ef151d660b817ccdfab3ece092a64780493ca2e636L312

**Comment by aqk** [comment]
Is the `softfork` extension argument mandatory?

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-2074b76a9ca8bff0a579d5ef151d660b817ccdfab3ece092a64780493ca2e636R340

**Comment by arvidn** [comment]
> Is the softfork extension argument mandatory?

technically it's not, in order to be backwards compatible. but we will require it in all our soft-forks, by convention.

**Comment by aqk** [comment]
For `_extension: OperatorSet`, is there a way to specify a combination of several OperatorSets?

Or were you thinking "just specify a complete set for each new dialect" ?

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-2074b76a9ca8bff0a579d5ef151d660b817ccdfab3ece092a64780493ca2e636R354

**Comment by prozacchiwawa** [comment]
regarding the extension set, apologies, that particular thing fell through the cracks.  pr here.
https://github.com/Chia-Network/clvm_tools_rs/pull/202

**Comment by prozacchiwawa** [comment]
added a comment to OPERATORS_LATEST_VERSION.

**Comment by prozacchiwawa** [comment]
much of the rest of the commentary here was about the clvmr upgrade, which went into a separate pr and got merged.

**Comment by aqk** [comment]
Line too long!!

https://github.com/Chia-Network/clvm_tools_rs/pull/180/files#diff-1831f1f42490a8fe847e2516a87bd681c15b6883489dffc8a597ef09eaffcae8R2

**Comment by arvidn** [comment]
> For `_extension: OperatorSet`, is there a way to specify a combination of several OperatorSets?
> 
> Or were you thinking "just specify a complete set for each new dialect" ?

At the core of it, we can do whatever we want, since anything is a soft-fork anyway. Regarding the CLVM implementation, and how it interprets this field, it's an enum that essentially specifies the soft-fork version. We don't expect to have a lot of soft-fork in flight in parallel, so a version is enough. We considered a bitmask, but that would limit us to 64 soft-forks (in practice, then we would need a wider type), without really providing any benefit, since later soft-forks are expected to include all features from previous ones.

**Comment by prozacchiwawa** [comment]
This was changed to pass through the extension we were given at the top.
Initially i was thinking: "we should provide a consistent set for compile
time", but realized that can change (and should change) according to the
user's --operators-version preference.  That change was in a separate pr.

**Review by aqk — DISMISSED** [review]

**Review by aqk — APPROVED** [review]

**Pull request merged**